qubes_config/widgets/utils.py

```diff
@@ -0,0 +1,124 @@
+# -*- encoding: utf8 -*-
+#
+# The Qubes OS Project, http://www.qubes-os.org
+#
+# Copyright (C) 2020 Marta Marczykowska-Górecka
+#                               <marmarta@invisiblethingslab.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation; either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+"""Qubes helper functions"""
+import contextlib
+import os
+import time
+
+import fcntl
+
+import qubesadmin
+import qubesadmin.exc
+import qubesadmin.vm
+
+from typing import Optional, Any, Dict
+
+APPVIEWER_LOCK = "/var/run/qubes/appviewer.lock"
+DATA = "/var/run/qubes/qubes-clipboard.bin"
+FROM = "/var/run/qubes/qubes-clipboard.bin.source"
+XEVENT = "/var/run/qubes/qubes-clipboard.bin.xevent"
+
+@contextlib.contextmanager
+def appviewer_lock():
+    """Used for purposes of copying to Qubes Global Clipboard"""
+    fd = os.open(APPVIEWER_LOCK, os.O_RDWR | os.O_CREAT, 0o0666)
+    try:
+        fcntl.flock(fd, fcntl.LOCK_EX)
+        yield
+    finally:
+        fcntl.flock(fd, fcntl.LOCK_UN)
+        os.close(fd)
+
+
+def copy_to_global_clipboard(text: str):
+    """Copy provided text to global clipboard"""
+    with appviewer_lock():
+        with open(DATA, "w", encoding='utf-8') as contents:
+            contents.write(text)
+        with open(FROM, "w", encoding='ascii') as source:
+            source.write("dom0")
+        with open(XEVENT, "w", encoding='ascii') as timestamp:
+            timestamp.write(time.ctime())
```

Comment:
This isn't "normal" time, it's X server event time. See how it's done in qui/clipboard.py.